### DISCUSSION TITLE
Getting an auth0 bearer token using login flow

### DISCUSSION THREAD

**Discussion by OysteinAmundsen** [artilleryio/artillery#986] [Artillery]
Hi.

I'm new to artillery, and want to setup a basic stress-test. But all my url's are behind an auth wall (using Auth0) and I want to try to get a valid token for the testing session so that my backend does not throw me into a 401-spiral. The artillery docs explains that I can hook up to some lifecycle events using `processor`, but does not explain which lifecycle hooks I have available. I've managed to figure out that there is a `beforeRequest` hook I can use, but this does not seem optimal as this will probably run before every request. My tokens have at least an hour validity...

**So the main question is;** how do I construct a processor hook which taps into the auth0 login flow, retreives a token which can then be stored in an environment variable (or some other local mechanism) for use as an Authorization Bearer token in future requests done by artillery?

OR if this is a bad pattern to follow, what is the best practice for urls behind auth-walls? I've already thought about loging in first and then copying the token to an environment variable and use that, but this makes the test a bit harder to use since it requires a manual step.

Any input is greatly appreciated.

**Comment by hassy** [comment]
hi @OysteinAmundsen 👋 welcome!

There's a couple of different options you can use.

You could log in as part of the main testing scenario itself by sending login payloads to the Auth0-hosted auth endpoint, and capture the token from the response with a `capture` block.

If the login flow is complex, it may make more sense to do it directly in Javascript. There's a `beforeScenario` hook which runs exactly once for each virtual user. You can set scenario variables in the hook function by adding entries to `userContext.vars` object, which are then available as scenario variables which may be used in templates like `"{{ myVar }}"`.

If you only need one token which will be shared by all virtual users, you could use a `before` block, which runs just once per test. It's a scenario of its own but any variables captured in it become available to all VUs. There's an example here:

https://github.com/artilleryio/artillery/issues/959#issuecomment-802704127

Hope this helps! This would actually make for a great example in [artillery-examples](https://github.com/artilleryio/artillery-examples/) repo.

**Reply by OysteinAmundsen** [reply]
Thank you! I will try this out.
